# 1. Business Understanding

1.1 Overview

1.2 Problem Statement

1.3 Business Objectives

1.4 Determining The Project Goals

1.5 Determining the project success criteria

# 2. Data Understanding

### Describing the data

#### Column Nmes and Description

* `InvoiceNo` - Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
* `StockCode` - Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
* `Description`  Product (item) name. Nominal.
* `Quantity` - The quantities of each product (item) per transaction. Numeric.
* `InvoiceDate` - Invoice Date and time. Numeric, the day and time when each transaction was generated.
* `UnitPrice` - Unit price. Numeric, Product price per unit in sterling.
* `CustomerID` - Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
* `Country` -Country name. Nominal, the name of the country where each customer resides.

### Importing Libraries

In [10]:
# Basic data manipultion and analysis
import pandas as pd
import numpy as np
# Data visualisation Libraries
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
from scipy import stats
# Machine learning models
from surprise import Reader, Dataset
from surprise.model_selection import train_test_split
from surprise.model_selection import cross_validate
from surprise.prediction_algorithms import SVD
from surprise.prediction_algorithms import KNNWithMeans, KNNBasic, KNNBaseline
from surprise.model_selection import GridSearchCV
from surprise.prediction_algorithms import knns
from surprise.similarities import cosine, msd, pearson
from surprise import accuracy
from sklearn.metrics.pairwise import cosine_similarity
from surprise.model_selection.validation import cross_validate
import warnings
warnings.filterwarnings('ignore')

In [11]:
# importing the data
df = pd.read_csv("C:/Users/Rfwel/Desktop/Moringa Course Materials/Phase 4/Phase 4 project/dsc-Recommender-system/data.csv", index_col=0)
df.head()

StockCode                          Description  Quantity  \
InvoiceNo                                                            
536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
536365        71053                  WHITE METAL LANTERN         6   
536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

              InvoiceDate  UnitPrice  CustomerID         Country  
InvoiceNo                                                         
536365     12/1/2010 8:26       2.55     17850.0  United Kingdom  
536365     12/1/2010 8:26       3.39     17850.0  United Kingdom  
536365     12/1/2010 8:26       2.75     17850.0  United Kingdom  
536365     12/1/2010 8:26       3.39     17850.0  United Kingdom  
536365     12/1/2010 8:26       3.39     17850.0  United Kingdom

2.1 Collecting Data

2.2 Describing the data


3. Data Preparation

3.1 Creating a new data frame

3.1.1 Splitting the date

3.1.2 Adding Total Cost Column

3.2 Cleaning the Data

3.2.1 Duplicates

3.2.2 Missing Values Treatment

3.2.3 Outliers

3.2.3.1 Unit Price

3.2.3.2 Quantity

3.3.Exploratory Data Analysis

3.3.1.Univariate Analysis

3.3.1.1. Product

3.3.1.2. Country

3.3.1.3. Months

3.3.1.4. Time of Day

3.3.1.5. Day of the Week

3.3.1.6 Unit Price

3.3.2.Bivariate Analysis

3.3.2.1 Total Amount vs Time of the day

3.3.2.2 Total Amount vs Day

3.3.2.3 Quantity vs Time day

3.3.2.4 Quantity vs Day of the week

3.3.2.5 Quantity vs Unit price

3.3.3.1 Feature Engineering

3.3.3.Pre-processing of Data

4. Modeling

4.1 Collaborative Filtering

4.1.1 Memory Based (User-User) - Cosine Similarity

4.1.2 Cosine Similarity between the Customers

4.1.3 Testing The Cosine Similarity

4.1.3.1 Top 5 Customers

4.1.3.2 Similar Customers to our Top Customer

4.1.3.3 Product Recommendation

4.1.2 Collaborative Filtering - Using Surprise Library

4.1.2.1 Creating a Class for Collaborative Filtering Models

4.1.2.1.1 Memory_Based Model using Cosine and Pearson Similarity


4.1.2.2 Model_Based Method - Singular Value Decomposition

Observations

Limitations

Recommendation